automation/routines.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2023 Searis AS
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package automation
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"slices"
+	"strings"
+)
+
+// Routines describe a set of named (sub-)routines. Routines can be nested by
+// letting the value be a Routines entry.
+//
+// For usability reasons, keys are recommended to only contain ASCII
+// alphanumerical characters (0-9, A-Z, a-z), dash (-) and underscore (_).
+//
+// Keys must not contain the slash (/) or asterisk (*) characters as they hold
+// special meaning during matching. The question mark (?) character should be
+// considered reserved. Keys must also not be empty strings. Failing to follow
+// these restrictions will result in undefined behavior for the SubRoutines
+// method.
+type Routines map[string]Routine
```

Comment:
Can we instead add `parsing/validation` logic of keys rather than hoping users will read this doc message?

Reply:
If we make the map a private struct field and add a `NewRoutines(map[string]Routine) Routines` method we could do validation. However, I want to avoid this, because I thing this:

```go
var routines = automation.Routines{
    "publish": automation.Routines{
        "routine1": publishRoutine1,
     } 
}
```

Is simpler to set-up than this:
```go
var routines = automation.NewRoutines(map[string]Routine{
    "publish": automation.NewRoutines(map[string]Routine{
        "routine1": publishRoutine1,
     }),
})
```